### PR TITLE
`VotingVanguardsRootstockCollective` NFT smart contract

### PR DESCRIPTION
**Pull request by cybervoid0**
# What

##  `VotingVanguardsRootstockCollective` NFT smart contract

TLDR:
- The contract fully (maybe almost) complies to  [the requirements from the ticked](https://rsklabs.atlassian.net/browse/DAO-861)
- A bunch of tests for different scenarios
- Deployment script
- Comprehensive test including gas fee measuring 

This pull request introduces a new smart contract for the Voting Vanguards Rootstock Collective, along with corresponding configuration, deployment, and testing updates. The most important changes include the creation of the `VotingVanguardsRootstockCollective` contract, the addition of a deployment module, parameter configuration for testnet, and comprehensive tests for the new contract.

### CODE REVIEW

File: contracts/NFT/VotingVanguardsRootstockCollective.sol [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+   * @dev Checks whether the specified address (`caller`) has voted on any of the last
+   * `numProposals` proposals. Iterates through the most recent proposals to determine
+   * if a vote exists for the caller.
+   *
+   * @param caller The address to check for voting activity.
+   * @return True if the caller has voted on at least one of the checked proposals, otherwise false.
+   */
+  function hasVoted(address caller) public view virtual returns (bool) {
+    uint256 firstCheckIndex = governor.proposalCount();
+    uint256 lastCheckIndex = firstCheckIndex > proposalCount ? firstCheckIndex - proposalCount : 0;
```
Comment (TravellerOnTheRun) [inline contracts/NFT/VotingVanguardsRootstockCollective.sol]: The variable name chosen for stored number of proposals to check confuses me, let's rename to `proposalsAmountToCheck`, `proposalsAmount`

File: contracts/NFT/VotingVanguardsRootstockCollective.sol [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  /**
+   * @dev Checks whether the specified address (`caller`) has voted on any of the last
+   * `numProposals` proposals. Iterates through the most recent proposals to determine
+   * if a vote exists for the caller.
+   *
+   * @param caller The address to check for voting activity.
+   * @return True if the caller has voted on at least one of the checked proposals, otherwise false.
+   */
+  function hasVoted(address caller) public view virtual returns (bool) {
+    uint256 firstCheckIndex = governor.proposalCount();
```
Comment (TravellerOnTheRun) [inline contracts/NFT/VotingVanguardsRootstockCollective.sol]: This is not the index, should it be  `- 1`?

File: contracts/NFT/VotingVanguardsRootstockCollective.sol [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    uint256 lastCheckIndex = firstCheckIndex > proposalCount ? firstCheckIndex - proposalCount : 0;
+    for (uint256 i = firstCheckIndex; i > lastCheckIndex; ) {
+      // slither-disable-next-line unused-return
+      (uint256 proposalId, , , , ) = governor.proposalDetailsAt(i - 1);
```
Comment (TravellerOnTheRun) [inline contracts/NFT/VotingVanguardsRootstockCollective.sol:97]: let's move that to avoid confusion to `firstCheckIndex`

File: test/Vanguard.test.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  let deployer: SignerWithAddress
+  let voter: SignerWithAddress
+  let proposalTarget: SignerWithAddress
+
+  before(async () => {
+    ;[deployer, voter, proposalTarget] = await ethers.getSigners()
+  })
+
+  // give voting power
+  const enfranchise = async (rif: RIFToken, stRIF: StRIFToken) => {
```
Comment (TravellerOnTheRun) [inline test/Vanguard.test.ts:37]: fancy word :D

File: test/Vanguard.test.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            ipfsFolderCid,
+          },
+        },
+      })
+    ).VanguardNFT as unknown as VotingVanguardsRootstockCollective
+    await enfranchise(contracts.rif, contracts.stRIF)
+    return { ...contracts, vanguard }
+  }
+
+  const createProposal = async (governor: GovernorRootstockCollective) => {
```
Comment (TravellerOnTheRun) [inline test/Vanguard.test.ts:74]: why did you not test on `TESTNET`? Hardhat network is not guaranteed to have the same amount of gas

File: test/Vanguard.test.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    before(async () => {
+      const contracts = await loadFixture(deploy)
+      governor = contracts.governor
+      vanguard = contracts.vanguard
+      stRif = contracts.stRIF
+    })
+
+    describe('Upon deployment', async () => {
+      it('Contract names should be set after the deployment', async () => {
+        expect(await governor.name()).to.equal('GovernorRootstockCollective')
```
Comment (TravellerOnTheRun) [inline test/Vanguard.test.ts:97]: Why do you redeploy the governor?

File: test/Vanguard.test.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+      it('StRif address should be set in the Vanguard contract', async () => {
+        expect(await vanguard.stRif()).to.equal(await stRif.getAddress())
+      })
+      it('Mint limit should be set', async () => {
+        expect(await vanguard.mintLimit()).to.equal(mintLimit)
+      })
+      it('All tokens should be available for minting', async () => {
+        expect(await vanguard.tokensAvailable()).to.equal(maxSupply)
+      })
+      it('Number of proposal to search should be set', async () => {
```
Comment (TravellerOnTheRun) [inline test/Vanguard.test.ts]: You have two of the same tests, look up `it('number of past votes to be checked should be set up'`

**Review by TravellerOnTheRun — CHANGES_REQUESTED** [review]

File: contracts/NFT/VotingVanguardsRootstockCollective.sol (hunk 2 above)
Comment (TravellerOnTheRun) [reply contracts/NFT/VotingVanguardsRootstockCollective.sol]: Still think it's not an index and hence you need to make one

**Review by TravellerOnTheRun — CHANGES_REQUESTED** [review]
Still one change left to avoid confusion and have a better code readability in the future

**Review by rodrigoncalves — APPROVED** [review]

**Review by TravellerOnTheRun — APPROVED** [review]
